packages/strapi-admin/services/permission/engine.js

```diff
@@ -179,48 +181,91 @@ module.exports = conditionProvider => {
       // 1. Replace each condition name by its associated value
       const resolveConditions = map(conditionProvider.get);
 
-      // 2. Only keep the handler of each condition
-      const pickHandlers = map(prop('handler'));
-
-      // 3. Filter conditions, only keep objects and functions
-      const filterValidConditions = filter(isObject);
+      // 2. Filter conditions, only keep objects and functions
+      const filterValidConditions = filter(condition => isObject(condition.handler));
 
-      // 4. Evaluate the conditions if they're a function, returns the object otherwise
+      // 3. Evaluate the conditions if they're a function, returns the object otherwise
       const evaluateConditions = conditions => {
         return Promise.all(
-          conditions.map(cond =>
-            isFunction(cond)
-              ? cond(user, merge(conditionOptions, { permission: cloneDeep(permission) }))
-              : cond
-          )
+          conditions.map(async cond => ({
+            ...cond,
+            handler: isFunction(cond.handler)
```

Comment:
Do we really support the handler to be something else than a function ? I think we shouldn't so the handler must be a function it would be simpler. Maybe we need to check the doc first to make sure we didn't document this

Reply:
We do accept function & objects + some documentation is available (https://strapi.io/documentation/developer-docs/latest/setup-deployment-guides/configurations.html#using-the-condition-handler)

Reply:
Ok that's too bad. the handler being an object is kind of ugly :) and inconsistent. We shoudl clean that in v4

Reply:
Follow-up:
As discussed internally and since the condition API is currently labelled as unstable, we'll proceed with the removal of object handler support.
I'll make a PR to the documentation & write a migration guide concerning this change.